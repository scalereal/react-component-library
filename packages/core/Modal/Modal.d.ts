import * as React from 'react';
import { ModalProps } from './types';
export declare const textSizes: ModalProps['titleSize'][];
export declare const descriptionSizes: ModalProps['descriptionSize'][];
export declare const Modal: React.FC<ModalProps>;
export default Modal;
