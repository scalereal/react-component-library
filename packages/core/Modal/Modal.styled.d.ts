import { ModalProps } from './types';
export declare const Modal: import("styled-components").StyledComponent<"div", any, ModalProps, never>;
export declare const ModalHeader: import("styled-components").StyledComponent<"div", any, ModalProps, never>;
export declare const ModalTitle: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const ModalBody: import("styled-components").StyledComponent<"div", any, ModalProps, never>;
export declare const ModalDescription: import("styled-components").StyledComponent<"div", any, {}, never>;
export declare const ModalFooter: import("styled-components").StyledComponent<"div", any, ModalProps, never>;
