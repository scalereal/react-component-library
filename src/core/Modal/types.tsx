import React, { RefObject } from 'react';
import { AnyStyledComponent } from 'styled-components';
import { TextSizes } from "../../theme/core/text/types"

export interface WithStyle {
    Style?: AnyStyledComponent;
}

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
export type HTMLPropsReact<T> = Omit<React.HTMLProps<T>, 'ref' | 'as'>;

export interface HTMLProps<T> extends HTMLPropsReact<T> {
    ref?: ((instance: T) => void) | RefObject<T>;
}
export interface WithThemeProp {
    // eslint-disable-next-line no-undef
    as?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
}
export interface ModalProps extends HTMLProps<HTMLDivElement>, WithThemeProp {
    /** Modal Title */
    title: string;
    /** Modal Description */
    description: string;
    /** Modal Border Color */
    borderColor?:string;
    /** Modal Shadow Color */
    shadowColor?:string;
    /** Modal Title Size */
    titleSize: TextSizes;
    /** Modal Description Size */
    descriptionSize: TextSizes;
    /** Modal Title Color */
    titleColor?:string;
    /** Modal Description Color */
    descriptionColor?:string;
};
