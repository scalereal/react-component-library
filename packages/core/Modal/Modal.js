import React from 'react';
import * as Styled from './Modal.styled';
import Text from '../Text/Text';
var Toast = function (_a) {
    var title = _a.title, description = _a.description, borderColor = _a.borderColor, shadowColor = _a.shadowColor, titleSize = _a.titleSize, titleColor = _a.titleColor, descriptionSize = _a.descriptionSize, descriptionColor = _a.descriptionColor;
    return (React.createElement(Styled.Modal, { title: title, description: description, shadowColor: shadowColor, borderColor: borderColor, titleSize: titleSize, descriptionSize: descriptionSize },
        React.createElement(Styled.ModalHeader, { title: title, description: description, shadowColor: shadowColor, borderColor: borderColor, titleSize: titleSize, descriptionSize: descriptionSize },
            React.createElement(Styled.ModalTitle, null,
                React.createElement(Text, { textSize: titleSize, lineHeight: "1.5", textColor: titleColor }, title))),
        React.createElement(Styled.ModalBody, { title: title, description: description, shadowColor: shadowColor, borderColor: borderColor, titleSize: titleSize, descriptionSize: descriptionSize },
            React.createElement(Styled.ModalDescription, null,
                React.createElement(Text, { textSize: descriptionSize, textColor: descriptionColor }, description))),
        React.createElement(Styled.ModalFooter, { title: title, description: description, shadowColor: shadowColor, borderColor: borderColor, titleSize: titleSize, descriptionSize: descriptionSize })));
};
export default Toast;
