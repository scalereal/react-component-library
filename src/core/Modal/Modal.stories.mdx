import Modal, { textSizes, descriptionSizes } from './Modal';
import { Preview, Story, Meta, Props, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import { withKnobs, color, boolean, select } from '@storybook/addon-knobs';

<Meta
    title="Components/Core/Modal"
    component={Modal}
    argTypes={{
        shadowColor: {
            control: 'color'
        },
        borderColor: {
            control: 'color'
        },
        titleColor: {
            control: 'color'
        },
        descriptionColor: {
            control: 'color'
        },
        titleSize: {
            control: {
                type: 'select',
                options: Object.values(textSizes)
            }
        },
        descriptionSize: {
            control: {
                type: 'select',
                options: Object.values(descriptionSizes)
            }
        }
    }}
/>

# Modal

Displays a Modal

export const Template = ({
    title,
    description,
    borderColor,
    shadowColor,
    titleSize,
    titleColor,
    descriptionSize,
    descriptionColor,
    ...args
}) => {
    return (
        <Modal
            title={title}
            description={description}
            shadowColor={shadowColor}
            borderColor={borderColor}
            titleSize={titleSize}
            descriptionSize={descriptionSize}
            titleColor={titleColor}
            descriptionColor={descriptionColor}
            {...args}
        />
    );
};

<Preview>
    <Story
        name="Modal"
        args={{
            title: 'Title',
            description: 'Lorem Ipsum is simply dummy text',
            titleSize: 'M',
            descriptionSize: 'XS'
        }}
    >
        {Template.bind({})}
    </Story>
</Preview>

### Color Variations

You can either change the theme or pass the colors as props.

<Preview withToolbar>
    <div style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Modal
            title="Title"
            description="Lorem Ipsum is simply dummy text"
            shadowColor="blue"
            borderColor="blue"
            titleSize="M"
            descriptionSize="XS"
            titleColor="blue"
            descriptionColor="green"
        />
        <Modal
            title="Title"
            description="Lorem Ipsum is simply dummy text"
            shadowColor="green"
            borderColor="green"
            titleSize="M"
            descriptionSize="XS"
            titleColor="green"
            descriptionColor="blue"
        />
    </div>
</Preview>

### Sizes

There are five sizes available `XS` | `S` | `M` | `L` | `XL`.

<Preview withToolbar>
    <div style={{ display: 'flex', flexDirection: 'column', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Modal title="Title" description="Lorem Ipsum is simply dummy text" titleSize="XS" descriptionSize="XS" />
        <Modal title="Title" description="Lorem Ipsum is simply dummy text" titleSize="S" descriptionSize="XS" />
        <Modal title="Title" description="Lorem Ipsum is simply dummy text" titleSize="M" descriptionSize="S" />
        <Modal title="Title" description="Lorem Ipsum is simply dummy text" titleSize="L" descriptionSize="M" />
        <Modal title="Title" description="Lorem Ipsum is simply dummy text" titleSize="XL" descriptionSize="L" />
    </div>
</Preview>

### Props

<Props story="Modal" />
