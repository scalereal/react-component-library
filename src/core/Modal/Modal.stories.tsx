import { Meta, Story } from '@storybook/react/types-6-0'
import React from 'react'
import Modal from './Modal'
import { ModalProps } from './types'

const textSizes : ModalProps['titleSize'][] = ['XS', 'S', 'M', 'L', 'XL']
const descriptionSizes : ModalProps['descriptionSize'][] = ['XS', 'S', 'M', 'L', 'XL']

export default {
    title: 'Components/Core/Modal',
    component: Modal,
    argTypes: {
        shadowColor:{
            control:'color'
        },
        borderColor:{
            control:'color'
        },
        titleColor:{
            control:'color'
        },
        descriptionColor:{
            control:'color'
        },
        titleSize:{
            control: {
                type: 'select',
                options: Object.values(textSizes),
            },
        },
        descriptionSize:{
            control: {
                type: 'select',
                options: Object.values(descriptionSizes),
            },
        }
    }
} as Meta

const Template:Story = ({ title, description,borderColor, shadowColor, titleSize,titleColor,descriptionSize,descriptionColor, ...args}) => {
    return <Modal title={title} description={description} shadowColor={shadowColor} borderColor={borderColor} titleSize={titleSize} descriptionSize={descriptionSize} titleColor={titleColor} descriptionColor={descriptionColor} {...args} />
}

export const Default = Template.bind({});
Default.args = { 
    title:'Title',
    description:'Lorem Ipsum is simply dummy text',
    titleSize:'M',
    descriptionSize:'XS',
}
