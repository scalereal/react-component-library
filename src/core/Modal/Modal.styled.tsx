import styled from 'styled-components';
import { ModalProps } from './types';
import { defaultTheme } from '../../theme';

export const Modal = styled('div')<ModalProps>`
    position: relative;
    display: flex;
    flex-direction: column;
    width: 100%;
    pointer-events: auto;
    background-color: #fff;
    background-clip: padding-box;
    border: 1px solid ${({ theme, borderColor }) => borderColor ? borderColor : theme.modal.defaults.borderColor};
    border-radius: .3rem;
    outline: 0;
    box-shadow: 0 0.5rem 2rem ${({ theme, shadowColor }) => shadowColor ? shadowColor : theme.modal.defaults.shadowColor};
`;
Modal.defaultProps = {
    theme:defaultTheme
}
export const ModalHeader = styled('div')<ModalProps>`
    display: flex;
    flex-shrink: 0;
    align-items: center;
    justify-content: space-between;
    padding: 1rem 1rem;
    border-bottom: 1px solid ${({ theme, borderColor }) => borderColor ? borderColor : theme.modal.defaults.borderColor};
    border-top-left-radius: calc(.3rem - 1px);
    border-top-right-radius: calc(.3rem - 1px);
`;
ModalHeader.defaultProps = {
    theme:defaultTheme
}
export const ModalTitle = styled('div')`
    margin-bottom: 0;
`;
export const ModalBody = styled('div')<ModalProps>`
    position: relative;
    flex: 1 1 auto;
    padding: 1rem;
`;
export const ModalDescription = styled('div')`
    margin-top: 0;
    margin-bottom: 1rem;
`;
export const ModalFooter = styled('div')<ModalProps>`
    display: flex;
    flex-wrap: wrap;
    flex-shrink: 0;
    align-items: center;
    justify-content: flex-end;
    padding: .75rem;
    border-top: 1px solid ${({ theme, borderColor }) => borderColor ? borderColor : theme.modal.defaults.borderColor};
    border-bottom-right-radius: calc(.3rem - 1px);
    border-bottom-left-radius: calc(.3rem - 1px);
`;
ModalFooter.defaultProps = {
    theme:defaultTheme
}
