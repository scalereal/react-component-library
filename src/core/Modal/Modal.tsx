import * as React from 'react'
import Text from '../Text/Text'
import * as Styled from './Modal.styled'
import { ModalProps } from './types'

export const textSizes : ModalProps['titleSize'][] = ['XS', 'S', 'M', 'L', 'XL']
export const descriptionSizes : ModalProps['descriptionSize'][] = ['XS', 'S', 'M', 'L', 'XL']

const Toast:React.FC<ModalProps> = ({ title, description, borderColor, shadowColor, titleSize,titleColor,descriptionSize,descriptionColor } : ModalProps) => {

    return (
        <Styled.Modal title={title} description={description} shadowColor={shadowColor} borderColor={borderColor} titleSize={titleSize} descriptionSize={descriptionSize}>
            <Styled.ModalHeader title={title} description={description} shadowColor={shadowColor} borderColor={borderColor} titleSize={titleSize} descriptionSize={descriptionSize}>
                <Styled.ModalTitle>
                     <Text textSize={titleSize} lineHeight="1.5" textColor={titleColor}>{title}</Text>
                </Styled.ModalTitle>
            </Styled.ModalHeader>
            <Styled.ModalBody title={title} description={description} shadowColor={shadowColor} borderColor={borderColor} titleSize={titleSize} descriptionSize={descriptionSize}>
              <Styled.ModalDescription>
                  <Text textSize={descriptionSize} textColor={descriptionColor}>{description}</Text>
              </Styled.ModalDescription>
            </Styled.ModalBody>
            <Styled.ModalFooter title={title} description={description} shadowColor={shadowColor} borderColor={borderColor} titleSize={titleSize} descriptionSize={descriptionSize}/>
        </Styled.Modal>
    )
}

export default Toast
