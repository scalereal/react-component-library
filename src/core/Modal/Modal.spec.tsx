import { render } from '@testing-library/react';
import * as React from 'react';
import Modal from "./Modal"


describe('Modal', () => {
    it('should render title and description text', () => {
        const { container } = render(<Modal  title="Title" description="Description" titleSize='S' descriptionSize='S'/>);
        expect(container).toMatchSnapshot();
    });
});
