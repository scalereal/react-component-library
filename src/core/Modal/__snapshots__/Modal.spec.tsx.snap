// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render title and description text 1`] = `
<div>
  <div
    class="sc-bdfBwQ kOUwLq"
    title="Title"
  >
    <div
      class="sc-gsTCUz kdZmDy"
      title="Title"
    >
      <div
        class="sc-dlfnbm eVErIe"
      >
        <span
          class="sc-gKsewC hWUpWw"
        >
          Title
        </span>
      </div>
    </div>
    <div
      class="sc-hKgILt iEmfbb"
      title="Title"
    >
      <div
        class="sc-eCssSg dkxmEH"
      >
        <span
          class="sc-gKsewC fXPGYv"
        >
          Description
        </span>
      </div>
    </div>
    <div
      class="sc-jSgupP fOOCZY"
      title="Title"
    />
  </div>
</div>
`;
